rlp/sedes/serializable.py

```diff
@@ -0,0 +1,299 @@
+            return False
+        return self.serialize(self) == other.serialize(other)
+
+    @classmethod
+    def get_sedes(cls):
+        return cls._meta.sedes
+
+    @classmethod
+    def serialize(cls, obj):
+        try:
+            result = cls.get_sedes().serialize(obj)
+        except ListSerializationError as e:
+            raise ObjectSerializationError(obj=obj, sedes=cls, list_exception=e)
+        else:
+            return result
+
+    @classmethod
+    def deserialize(cls, serial, exclude=None, mutable=False, **kwargs):
+        try:
+            values = cls.get_sedes().deserialize(serial)
+        except ListDeserializationError as e:
+            raise ObjectDeserializationError(serial=serial, sedes=cls, list_exception=e)
+
+        params = {
+            field: value
+            for field, value
+            in zip(cls._meta.field_names, values)
+            if exclude is None or field not in exclude
+        }
+        obj = cls(**dict(list(params.items()) + list(kwargs.items())))
+
+        # TODO: mutability
+        if mutable:
+            return obj.as_mutable()
+        else:
+            return obj.as_immutable()
+
+    @classmethod
+    def exclude(cls, excluded_fields):
+        """Create a new sedes considering only a reduced set of fields."""
+        raise NotImplementedError("Not yet implemented")
+
+    @property
+    def is_immutable(self):
+        return not self.is_mutable
+
+    #
+    # Abstract methods
+    #
+    @abc.abstractmethod
+    def as_immutable(self):
+        pass
+
+    @abc.abstractmethod
+    def as_mutable(self):
+        pass
+
+    @abc.abstractmethod
+    def _get_field_attrs(self):
+        pass
+
+
+def make_immutable(value):
+    if hasattr(value, 'as_immutable'):
+        return value.as_immutable()
+    elif not isinstance(value, (bytes, str)) and isinstance(value, collections.Sequence):
+        return tuple(make_immutable(item) for item in value)
+    else:
+        return value
+
+
+def make_mutable(value):
+    if hasattr(value, 'as_mutable'):
+        return value.as_mutable()
+    elif not isinstance(value, (bytes, str)) and isinstance(value, collections.Sequence):
+        return list(make_mutable(item) for item in value)
+    else:
+        return value
+
+
+class ImmutableMixin:
+    is_mutable = False
+
+    _cached_rlp_value = None
+
+    @property
+    def _cached_rlp(self):
+        return self._cached_rlp_value
+
+    @_cached_rlp.setter
+    def _cached_rlp(self, value):
+        self._cached_rlp_value = value
+
+    def as_immutable(self):
+        return self
+
+    def as_mutable(self):
+        return self._meta.mutable_cls(*(
+            make_mutable(arg)
+            for arg
+            in self
+        ))
+
+    def _get_field_attrs(self):
+        return self._meta.field_attrs
+
+
+class WritableSerializable(BaseSerializable):
+    is_mutable = True
+
+    _cached_rlp_value = None
+    _cached_rlp_key = None
```

Comment:
There's something about having two separate attributes for this that makes me uncomfortable... Did you consider using a dict for that? Maybe `_cached_snapshot = {}`, or maybe a `_get_cached_rlp(snapshot_key)` method decorated with a `lru_cache`

Reply:
Alright, I've changed to use a class level cache.  Each subclass of `Serializable` now has it's own cache which is shared across instances of those objects as well as across both the mutable and immutable versions of those objects.